### PR TITLE
more quic servers

### PR DESCRIPTION
**Pull request by segfaultdoc**

### CODE REVIEW

File: docker-compose.yaml [hunk 1]
```diff
@@ -13,10 +13,10 @@ services:
       - KEYPAIR_PATH=/etc/solana/id.json
       - PUBLIC_IP=127.0.0.1
       - TPU_BIND_IP=0.0.0.0
-      - TPU_PORT=10000
-      - TPU_FWD_PORT=10001
       - TPU_QUIC_PORT=10006
-      - TPU_QUIC_FWD_PORT=10007
+      - NUM_TPU_QUIC_SERVERS=10
```
Comment (esemeniuc) [inline docker-compose.yaml:17]: `NUM_TPU_QUIC_LISTENERS` ?

File: transaction-relayer/src/main.rs [hunk 2]
```diff
@@ -58,23 +59,29 @@ static GLOBAL: Jemalloc = Jemalloc;
 #[derive(Parser, Debug)]
 #[clap(author, version, about, long_about = None)]
 struct Args {
-    /// Port to bind to advertise for TPU
-    /// NOTE: There is no longer a socket created at this port since UDP transaction receiving is
-    /// deprecated.
-    #[arg(long, env, default_value_t = 11_222)]
+    /// DEPRECATED, will be removed in a future release.
```
Comment (esemeniuc) [inline transaction-relayer/src/main.rs:62]: can we use something to mark as deprecated? ie https://github.com/rust-lang/rust/issues/102777

File: transaction-relayer/src/main.rs [hunk 3]
```diff
@@ -210,29 +217,64 @@ struct Sockets {
...
-    .expect("to bind tpu_quic sockets");
+    for tpu_port in tpu_ports.start..tpu_ports.end {
+        assert!(!tpu_fwd_ports.contains(&tpu_port));
+    }
+
+    let (tpu_p, tpu_quic_sockets): (Vec<_>, Vec<_>) = (0..args.num_tpu_quic_servers)
+        .map(|i| {
+            let (port, mut sock) = multi_bind_in_range(
+                IpAddr::V4(Ipv4Addr::from([0, 0, 0, 0])),
+                (tpu_ports.start + i as u16, tpu_ports.start + 1 + i as u16),
```
Comment (esemeniuc) [inline transaction-relayer/src/main.rs:246]: can do `(0..args.num_tpu_quic_servers as u16)` and avoid casting here

File: transaction-relayer/src/main.rs [hunk 4]
```diff
@@ -58,23 +59,29 @@ static GLOBAL: Jemalloc = Jemalloc;
...
-    /// deprecated.
-    #[arg(long, env, default_value_t = 11_223)]
+    /// DEPRECATED, will be removed in a future release.
+    #[arg(long, env, default_value_t = 0)]
     tpu_fwd_port: u16,
 
-    /// Port to bind to for tpu packets. Needs to be tpu_port + 6
+    /// Number of tpu quic servers to spawn.
+    #[arg(long, env, default_value_t = 1)]
+    num_tpu_quic_servers: usize,
```
Comment (esemeniuc) [inline transaction-relayer/src/main.rs:72]: why not make u16 and avoid all the casting?

Comment (esemeniuc) [reply transaction-relayer/src/main.rs:72]: clap will error if you try a value larger than u16::MAX

```
error: invalid value '999999999999' for '--val <VAL>': 999999999999 is not in 0..=65535

For more information, try '--help'.

Process finished with exit code 2
```

File: transaction-relayer/src/main.rs [hunk 5]
```diff
@@ -411,8 +453,9 @@ fn main() {
         delay_packet_receiver,
         leader_cache.handle(),
         public_ip,
-        args.tpu_port,
-        args.tpu_fwd_port,
+        (args.tpu_quic_port..args.tpu_quic_port + args.num_tpu_quic_servers as u16).collect(),
```
Comment (esemeniuc) [inline transaction-relayer/src/main.rs:456]: use range instead?

File: transaction-relayer/src/main.rs [hunk 6]
```diff
@@ -411,8 +453,9 @@ fn main() {
...
+        (args.tpu_quic_fwd_port..args.tpu_quic_fwd_port + args.num_tpu_fwd_quic_servers as u16)
```
Comment (esemeniuc) [inline transaction-relayer/src/main.rs:457]: ditto

File: transaction-relayer/src/main.rs [hunk 7]
```diff
@@ -58,23 +59,29 @@ static GLOBAL: Jemalloc = Jemalloc;
...
-    /// Port to bind to for tpu packets. Needs to be tpu_port + 6
+    /// Number of tpu quic servers to spawn.
+    #[arg(long, env, default_value_t = 1)]
+    num_tpu_quic_servers: usize,
+
+    /// Number of tpu fwd quic servers to spawn.
+    #[arg(long, env, default_value_t = 1)]
+    num_tpu_fwd_quic_servers: usize,
+
+    /// Port to bind to for tpu packets. Need to return port - 6 to validators.
```
Comment (esemeniuc) [inline transaction-relayer/src/main.rs:78]: why `port - 6`? some comments would be very helpful

File: transaction-relayer/src/main.rs (hunk 5 above)
Comment (jedleggett) [reply transaction-relayer/src/main.rs:456]: this syntax is an alias for a Range{}

File: transaction-relayer/src/main.rs [hunk 8]
```diff
@@ -59,26 +60,37 @@ static GLOBAL: Jemalloc = Jemalloc;
...
+    /// DEPRECATED, will be removed in a future release.
+    #[deprecated(since = "0.1.8", note = "UDP TPU_FWD disabled")]
+    #[arg(long, env, default_value_t = 0)]
     tpu_fwd_port: u16,
 
-    /// Port to bind to for tpu packets. Needs to be tpu_port + 6
+    /// Port to bind to for tpu quic packets.
+    /// Note: get_tpu_configs will return ths port - 6 to validators to match old UDP TPU definition.
+    /// Make sure to not overlap any tpu forward ports with the normal tpu ports.
+    /// The TPU will bind to all ports in the range of (tpu_quic_port, tpu_quic_port + num_tpu_quic_servers).
```
Comment (buffalu) [inline transaction-relayer/src/main.rs:76]: can add a comment for opening this port on firewall?

File: transaction-relayer/src/main.rs [hunk 9]
```diff
@@ -220,29 +232,61 @@ struct Sockets {
...
+    };
 
-    let (tpu_fwd_quic_bind_port, mut tpu_fwd_quic_sockets) = multi_bind_in_range(
-        IpAddr::V4(Ipv4Addr::from([0, 0, 0, 0])),
-        (args.tpu_quic_fwd_port, args.tpu_quic_fwd_port + 1),
-        1,
-    )
-    .expect("to bind tpu_quic sockets");
+    for tpu_port in tpu_ports.start..tpu_ports.end {
+        assert!(!tpu_fwd_ports.contains(&tpu_port));
```
Comment (buffalu) [inline transaction-relayer/src/main.rs:254]: should add more assertions here (striping comment)

Comment (jedleggett) [reply transaction-relayer/src/main.rs:254]: my understanding is that this is unnecessary after slack discussion.  

as long as the tpu_quic and tpu_fwd_quic ports don't overlap, this should be sufficient.

**Review by jedleggett — APPROVED** [review]

**Pull request merged**